plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusableConnectionFactory.java

```diff
@@ -0,0 +1,184 @@
+    private final ConnectionFactory delegate;
+
+    @Inject
+    public ReusableConnectionFactory(@ForReusableConnectionFactory ConnectionFactory delegate)
+    {
+        this(delegate, Duration.ofSeconds(2), 10);
+    }
+
+    ReusableConnectionFactory(ConnectionFactory delegate, Duration duration, long maximumSize)
+    {
+        this.connections = createConnectionsCache(duration, maximumSize);
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    // CacheBuilder.build(CacheLoader) is forbidden, because it does not support eviction for ongoing loads.
+    // In this class, loading is not used and cache is used more as a map. So this is safe.
+    @SuppressModernizer
+    private static Cache<String, Connection> createConnectionsCache(Duration duration, long maximumSize)
+    {
+        requireNonNull(duration, "duration is null");
+        return CacheBuilder.newBuilder()
+                .maximumSize(maximumSize)
+                .expireAfterWrite(duration)
+                .removalListener(ReusableConnectionFactory::onRemoval)
+                .build();
+    }
+
+    private static void onRemoval(RemovalNotification<String, Connection> notification)
+    {
+        if (notification.getCause() == EXPLICIT) {
+            // connection was taken from the cache
+            return;
+        }
+        try {
+            requireNonNull(notification.getValue(), "notification.getValue() is null");
+            notification.getValue().close();
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.asMap().remove(queryId);
+        if (connection != null) {
+            return connection;
+        }
+        return delegate.openConnection(session);
+    }
+
+    @Override
+    public void beginQuery(ConnectorSession session) {}
+
+    @Override
+    public void cleanupQuery(ConnectorSession session)
+    {
+        Connection connection = connections.asMap().remove(session.getQueryId());
+        if (connection != null) {
+            try {
+                connection.close();
+            }
+            catch (SQLException e) {
+                throw new TrinoException(JDBC_ERROR, e);
+            }
+        }
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        for (Connection connection : connections.asMap().values()) {
+            connection.close();
+        }
+        connections.invalidateAll();
+        delegate.close();
```

Comment:
I don't think we need to be so thread safe here. When somebody is closing a connection factory then they should do it gratefully, otherwise they risk that opening connection will fail. Also notice that we close connection factory only in tests.

Reply:
I was thinking about when dynamic catalogs arrive. But yes, we can worry about that when it actually arrives. Also might be simple fix by making the factory not a singleton and instead be once per catalog. Nothing to do here.